cas-server-support-yubikey/src/main/java/org/jasig/cas/adaptors/yubikey/YubiKeyAuthenticationHandler.java

```diff
@@ -0,0 +1,146 @@
+     * all YubiKey accounts are allowed to authenticate.
+     * <p/>
+     * <strong>WARNING: THIS CONSTRUCTOR RESULTS IN A
+     * CONFIGURATION THAT CONSIDERS ALL Yubikeys VALID FOR ALL USERS. YOU MUST NOT USE
+     * THIS CONSTRUCTOR IN PRODUCTION.</strong>
+     *
+     * @param clientId the client id
+     * @param secretKey the secret key
+     */
+    public YubiKeyAuthenticationHandler(@NotNull final Integer clientId,
+                                        @NotNull final String secretKey) {
+        this.client = YubicoClient.getClient(clientId, secretKey);
+    }
+
+    /**
+     * Prepares the Yubico client with the received clientId and secretKey. If you wish to
+     * limit the usage of this handler only to a particular set of yubikey accounts for a special
+     * group of users, you may provide an compliant implementation of {@link YubiKeyAccountRegistry}.
+     * By default, all accounts are allowed.
+     *
+     * @param clientId the client id
+     * @param secretKey the secret key
+     * @param registry the registry
+     */
+    public YubiKeyAuthenticationHandler(@NotNull final Integer clientId, final String secretKey,
+                                        @NotNull final YubiKeyAccountRegistry registry) {
+        this(clientId, secretKey);
+        this.registry = registry;
+    }
+
+    @Override
+    @PostConstruct
+    public void afterPropertiesSet() throws Exception {
+        if (this.registry == null) {
+            logger.warn("No YubiKey account registry is defined. All credentials are considered"
+                    + "eligible for YubiKey authentication. Consider providing an account registry via [{}]",
+                    YubiKeyAccountRegistry.class.getName());
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p/>
+     * Attempts to authenticate the received credentials using the Yubico cloud validation platform.
+     * In this implementation, the {@link UsernamePasswordCredential#getUsername()}
+     * is mapped to the <code>uid</code> which will be used by the plugged-in instance of the
+     * {@link YubiKeyAccountRegistry}
+     * and the {@link UsernamePasswordCredential#getPassword()} is the received
+     * one-time password token issued by the YubiKey device.
+     */
+    @Override
+    protected HandlerResult authenticateUsernamePasswordInternal(final UsernamePasswordCredential transformedCredential)
+            throws GeneralSecurityException, PreventedException {
+
+        final String uid = transformedCredential.getUsername();
+        final String otp = transformedCredential.getPassword();
+
+        if (!YubicoClient.isValidOTPFormat(otp)) {
+            logger.debug("Invalid OTP format [{}]", otp);
+            throw new FailedLoginException("OTP format is invalid");
+        }
+
+        final String publicId = YubicoClient.getPublicId(otp);
+        if (this.registry != null
+              &&!this.registry.isYubiKeyRegisteredFor(uid, publicId)) {
```

Comment:
The registry is used to authorize which users are actually allowed to use Yubikey. A separate registration process may be available that allows users to register their device with their userid and only then they can use YubiKey with a registered device. Otherwise, I could just log in with my own Yubikey as you! Just because the OTP is a valid token, it doesnt mean it's actually authorized and is a valid credential.